server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java

```diff
@@ -378,6 +384,36 @@ public void testPropagateDynamicWithDynamicMapper() throws Exception {
         assertNotNull(doc.rootDoc().getField("foo.bar.baz"));
     }
 
+    public void testPropagateDynamicRuntimeWithDynamicMapper() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(topMapping(b -> {
+            b.field("dynamic", false);
+            b.startObject("properties");
+            {
+                b.startObject("foo");
+                {
+                    b.field("type", "object");
+                    b.field("dynamic", "runtime");
+                    b.startObject("properties").endObject();
+                }
+                b.endObject();
+            }
+            b.endObject();
+        }));
+        ParsedDocument doc = mapper.parse(source(b -> {
+            b.startObject("foo");
+            {
+                b.field("baz", "test");
+                b.startObject("bar").field("baz", "something").endObject();
+            }
+            b.endObject();
+        }));
+        assertNull(doc.rootDoc().getField("foo.bar.baz"));
+        assertEquals("{\"_doc\":{\"dynamic\":\"false\"," +
+            "\"runtime\":{\"foo.bar.baz\":{\"type\":\"string\"},\"foo.baz\":{\"type\":\"string\"}}," +
+            "\"properties\":{\"foo\":{\"dynamic\":\"runtime\",\"properties\":{\"bar\":{\"type\":\"object\"}}}}}}",
```

Comment:
Are we happy that this intermediate object with no concrete leaf fields gets added?  It feels a bit weird to me, but I can see how it ends up being necessary because of all the book-keeping we do while we create dynamic mappings during parsing.

Reply:
good question: I have wondered the same, and wondered also: what do we do when a document is sent which contains no leaf fields, only objects? Effectively no lucene fields (besides metadata fields) are created, yet we map the objects. I followed the same pattern here, which kind of fits in the runtime section design, as it is applied on top of mappings, hence it is natural that it does not hold objects, and that only the leaf fields get mapped in there. As a consequence, unseen objects get dynamically mapped under properties in dynamic runtime mode. I can see how this may feel weird at first glance, though I wonder what the alternatives are, and what the downsides could be of this approach.

Reply:
replying to myself 6 months later: this would be the main downside of dynamically mapping objects under `dynamic:runtime`: https://github.com/elastic/elasticsearch/issues/70268 . I am thinking now that it makes more sense to skip mapping objects in dynamic runtime mode.